### PR TITLE
Scope deprecation warnings

### PR DESCRIPTION
**Pull request by comp-phys-marc**
**Context:** A deprecation warning is now raised when PL is imported, due to a recent change. We need to scope this warning appropriately and deal with it in the appropriate places.

**Description of the Change:** Scopes calls to `create_expand_fn`. Unfortunately, this means that warnings which would have been raised only on import before are now raised whenever a number of functions are called, revealing new places where we need to handle dep warnings or update functionality in the codebase.

**Benefits:** This is a pervasive warning that could be really distracting to users. We are getting rid of that confusing behaviour.

**Possible Drawbacks:** N/A

**Related Shortcut Stories:** [sc-107783]

### CODE REVIEW

File: pennylane/gradients/parameter_shift_cv.py [hunk 1]
```diff
@@ -504,6 +505,14 @@ def processing_fn(results):
     return gradient_tapes, processing_fn
 
 
+def _stop_at_expand_invalid_trainable(obj):
+    return (
+        isinstance(obj, MeasurementProcess)
+        or not any(math.requires_grad(d) for d in obj.data)
+        or obj.grad_method is not None
+    )
```
Comment (mudit2812) [inline pennylane/gradients/parameter_shift_cv.py]: We don't need to specify `isinstance(obj, MeasurementProcess)`. That condition will never be true because `decompose` only inspects `tape.operations`. Same comment for all other stopping conditions added in this PR
```suggestion
    return (
        not any(math.requires_grad(d) for d in obj.data)
        or obj.grad_method is not None
    )
```

File: pennylane/transforms/tape_expand.py [hunk 2]
```diff
@@ -180,11 +183,13 @@ def _trainable_multipar_stopping_fn(obj):
...
-    stop_at=_trainable_multipar_stopping_fn,
-    docstring=_expand_trainable_multipar_doc,
-)
+# pylint: disable=missing-function-docstring
+def expand_trainable_multipar(*args, **kwargs):
+    return create_expand_fn(
+        depth=None,
+        stop_at=_trainable_multipar_stopping_fn,
+        docstring=_expand_trainable_multipar_doc,
+    )(*args, **kwargs)
```
Comment (mudit2812) [inline pennylane/transforms/tape_expand.py:192]: Why can't we reimplement these using `transforms.decompose` rather than using deprecated code? That would allow us to continue using these gradient specific expansion functions in the `gradients` module rather than having to manually implement them for each gradient method

File: tests/transforms/test_tape_expand.py [hunk 3]
```diff
@@ -131,6 +131,36 @@ def test_device_only_expansion(self):
...
+
+        with qml.queuing.AnnotatedQueue() as q:
+            qml.RX(1.5, wires=0)
+            qml.Rot(-2.1, 1.0, -0.418, wires=1)
+            qml.Rot(-2.1, theta, -0.418, wires=1)
+            _CRX(1.5, wires=[0, 2])
+
+        tape = qml.tape.QuantumScript.from_queue(q)
+        with pytest.warns(PennyLaneDeprecationWarning, match="expand"):
+            new_tape = qml.transforms.expand_trainable_multipar(tape)
```
Comment (mudit2812) [inline tests/transforms/test_tape_expand.py:158]: Are `expand_trainable_multipar` and other similar expansion functions being deprecated? If not, then I don't think it's okay to have a deprecation warning be raised when they're used.

File: pennylane/transforms/tape_expand.py (hunk 2 above)
Comment (comp-phys-marc) [reply pennylane/transforms/tape_expand.py:192]: Hi @mudit2812 all the calls to `expand_trainable_multipar`, etc. have been replaced with calls to `decompose`. We are planning to remove this entire file.

File: tests/transforms/test_tape_expand.py (hunk 3 above)
Comment (comp-phys-marc) [reply tests/transforms/test_tape_expand.py:158]: We are removing them when we remove tape.expand.

File: pennylane/fourier/qnode_spectrum.py [hunk 4]
```diff
@@ -154,6 +159,18 @@ def _process_ids(
     return encoding_args, argnum
 
 
+def _multipar_stopping_fn(obj):
+    try:
+        return (
+            isinstance(obj, MeasurementProcess)
+            or len(obj.data) == 0
+            or (obj.has_generator and len(obj.generator().terms()[0]) == 1)
```
Comment (comp-phys-marc) [inline pennylane/fourier/qnode_spectrum.py]:
```suggestion
            len(obj.data) == 0
            or (obj.has_generator and len(obj.generator().terms()[0]) == 1)
```

File: pennylane/gradients/metric_tensor.py [hunk 5]
```diff
@@ -72,6 +74,25 @@ def _contract_metric_tensor_with_cjac(mt, cjac, tape):  # pylint: disable=unused
     return _mt_cjac_tdot(mt, cjac)
 
 
+def _multipar_stopping_fn(obj):
+    try:
+        return (
+            isinstance(obj, MeasurementProcess)
+            or len(obj.data) == 0
+            or (obj.has_generator and len(obj.generator().terms()[0]) == 1)
```
Comment (comp-phys-marc) [inline pennylane/gradients/metric_tensor.py]:
```suggestion
            len(obj.data) == 0
            or (obj.has_generator and len(obj.generator().terms()[0]) == 1)
```

File: pennylane/gradients/metric_tensor.py [hunk 6]
```diff
@@ -72,6 +74,25 @@ def _contract_metric_tensor_with_cjac(mt, cjac, tape):  # pylint: disable=unused
...
+        )
+    except TermsUndefinedError:
+        return True
+
+
+def _expand_nonunitary_gen_stop_at(obj):
+    return (
+        isinstance(obj, MeasurementProcess)
+        or len(obj.data) == 0
+        or (obj.has_generator and obj in has_unitary_generator)
```
Comment (comp-phys-marc) [inline pennylane/gradients/metric_tensor.py]:
```suggestion
        len(obj.data) == 0
        or (obj.has_generator and obj in has_unitary_generator)
```

File: pennylane/gradients/spsa_gradient.py [hunk 7]
```diff
@@ -61,6 +63,14 @@ def _rademacher_sampler(indices, num_params, *args, rng):
     return direction
 
 
+def _stop_at_expand_invalid_trainable(obj):
+    return (
+        isinstance(obj, MeasurementProcess)
+        or not any(math.requires_grad(d) for d in obj.data)
+        or obj.grad_method is not None
```
Comment (comp-phys-marc) [inline pennylane/gradients/spsa_gradient.py]:
```suggestion
        not any(math.requires_grad(d) for d in obj.data)
        or obj.grad_method is not None
```

File: pennylane/transforms/tape_expand.py (hunk 2 above)
Comment (comp-phys-marc) [reply pennylane/transforms/tape_expand.py:192]: I actually proposed the same thing you are saying on slack. @albi3ro was of the opinion that having a library of stopping conditions here (instead of moving them out of the gradients module) was not worth it since the stopping conditions are so simple.

File: tests/transforms/test_tape_expand.py (hunk 3 above)
Comment (comp-phys-marc) [reply tests/transforms/test_tape_expand.py:158]: Since they are literally just calls to `create_expand_fn` with certain stopping condtions, they are not worth having their own unique deprecation warnings.

File: pennylane/transforms/tape_expand.py (hunk 2 above)
Comment (mudit2812) [reply pennylane/transforms/tape_expand.py:192]: Thanks for the context. Based on this response and the response to my comment on testing, resolving.

**Review by mudit2812 — APPROVED** [review]

**Review by andrijapau — APPROVED** [review]
Thanks 🚀

**Review by gabrielasd — APPROVED** [review]
